src/iamserver/publicmessagehandler.hpp

```diff
@@ -0,0 +1,224 @@
+     * @param identHandler identification handler.
+     * @param permHandler permission handler.
+     * @param nodeInfoProvider node info provider.
+     * @param nodeManager node manager.
+     * @param provisionManager provision manager.
+     */
+    aos::Error Init(NodeController& nodeController, aos::iam::identhandler::IdentHandlerItf& identHandler,
+        aos::iam::permhandler::PermHandlerItf& permHandler, aos::iam::NodeInfoProviderItf& nodeInfoProvider,
+        aos::iam::nodemanager::NodeManagerItf&           nodeManager,
+        aos::iam::provisionmanager::ProvisionManagerItf& provisionManager);
+
+    /**
+     * Registers grpc services.
+     *
+     * @param builder server builder.
+     */
+    void RegisterServices(grpc::ServerBuilder& builder);
+
+    /**
+     * Node info change notification.
+     *
+     * @param info node info.
+     */
+    void OnNodeInfoChange(const aos::NodeInfo& info) override;
+
+    /**
+     * Node info removed notification.
+     *
+     * @param id id of the node been removed.
+     */
+    void OnNodeRemoved(const aos::String& id) override;
+
+    /**
+     * Subjects observer interface implementation.
+     *
+     * @param[in] messages subject changed messages.
+     * @returns Error.
+     */
+    aos::Error SubjectsChanged(const aos::Array<aos::StaticString<aos::cSubjectIDLen>>& messages) override;
+
+    /**
+     * Closes public message handler.
+     */
+    void Close();
+
+protected:
+    /**
+     * Server writer controller handles server writer streams.
+     */
+    template <typename T>
+    class ServerWriterController {
+    public:
+        /**
+         * Closes all streams.
+         */
+        void Close()
+        {
+            mIsRunning = false;
+
+            {
+                std::unique_lock lock {mMutex};
+
+                mLastMessage.reset();
+            }
+
+            mCV.notify_all();
+        }
+
+        /**
+         * Writes notification message to all streams.
+         *
+         * @param message notification message.
+         */
+        void WriteToSteams(const T& message)
+        {
+            {
+                std::unique_lock lock {mMutex};
+
+                mLastMessage = message;
+            }
+
+            mCV.notify_all();
+        }
+
+        /**
+         * Handles stream. Blocks the caller until the stream is closed.
+         *
+         * @param context server context.
+         * @param writer server writer.
+         * @return grpc::Status.
+         */
+        grpc::Status HandleStream(grpc::ServerContext* context, grpc::ServerWriter<T>* writer)
+        {
+            while (mIsRunning && !context->IsCancelled()) {
+                std::shared_lock lock {mMutex};
+
+                if (mCV.wait_for(lock, cWaitTimeout, [this] { return mLastMessage.has_value(); })) {
+                    // got notification, send it to the client
+                    if (!writer->Write(*mLastMessage)) {
```

Comment:
should we reset `mLastMessage` after we write a message? if it optional then wait_for will always pass, but in this case we will send the same message